libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp

```diff
@@ -1274,63 +1670,620 @@ void PeerConnectionImpl::OnLocalDescCreated(
+        // description, and update the state of the associated wrappers.
+        if (IsUnifiedPlan()) {
+          // Inspect transceiver directions, check for changes to update the
+          // interop layer with the actually negotiated direction.
+          std::vector<rtc::scoped_refptr<webrtc::RtpTransceiverInterface>>
+              changed_transceivers;
+          int mline_index = 0;  // native transceivers are in mline_index order
+          for (auto&& tr : peer_->GetTransceivers()) {
+            // If transceiver is created from the result of applying a local
+            // description, then the transceiver name is extracted from the
+            // sender, as the name should have been set by the user.
+            std::string name = ExtractTransceiverNameFromSender(tr->sender());
+            ErrorOr<RefPtr<Transceiver>> err =
+                GetOrCreateTransceiver(mline_index, tr, std::move(name));
+            RTC_DCHECK(err.ok());
+            err.value()->OnSessionDescUpdated(/*remote=*/false);
+            ++mline_index;
+          }
+        } else {
+          RTC_DCHECK(IsPlanB());
+          // Senders are already created during |CreateOffer()|, and receivers
+          // won't be created until the answer is received, so there is nothing
+          // to do here.
+        }
+
+        // Fire interop callback, if any
+        {
+          auto lock = std::scoped_lock{local_sdp_ready_to_send_callback_mutex_};
+          if (auto cb = local_sdp_ready_to_send_callback_) {
+            auto desc = peer_->local_description();
+            std::string type{SdpTypeToString(desc->GetType())};
+            std::string sdp;
+            desc->ToString(&sdp);
+            cb(type.c_str(), sdp.c_str());
+          }
+        }
+      });
   // SetLocalDescription will invoke observer.OnSuccess() once done, which
   // will in turn invoke the |local_sdp_ready_to_send_callback_| registered if
   // any, or do nothing otherwise. The observer is a mandatory parameter.
   peer_->SetLocalDescription(observer, desc);
 }
 
+void PeerConnectionImpl::OnLocalTrackAddedToAudioTransceiver(
+    Transceiver& transceiver,
+    LocalAudioTrack& track) {
+  rtc::CritScope lock(&tracks_mutex_);
+  RTC_DCHECK(std::find_if(transceivers_.begin(), transceivers_.end(),
+                          [&transceiver](const RefPtr<Transceiver>& tr) {
+                            return (
+                                (tr.get() == &transceiver) &&
+                                (tr->GetMediaKind() == mrsMediaKind::kAudio));
+                          }) != transceivers_.end());
+  RTC_DCHECK(std::find_if(local_audio_tracks_.begin(),
+                          local_audio_tracks_.end(),
+                          [&track](const RefPtr<LocalAudioTrack>& tr) {
+                            return (tr.get() == &track);
+                          }) == local_audio_tracks_.end());
+  local_audio_tracks_.push_back(&track);
```

Comment:
nit: it seems that we need a lot of work to keep these tracks array up to date. Can we do without? Any reason why we can't get to all the tracks by looking into the transceivers? We only seem to be enumerating them in the Close() method anyway.

Reply:
Actually we don't need to keep a reference to local tracks, and this is even useless and partly wrong since those objects are owned by the user, so should only be strongly referenced by the transceiver they are attached to, if any. And remote tracks are owned by transceivers, so it's a similar story, and can be removed too.

Fixed. That deleted a bunch of code :wink: